url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ARMv5.java

```diff
@@ -0,0 +1,958 @@
+        int pcOrg, cpsrOrg;
+
+        System.out.printf("Exception: Undefined instruction by '%s'.\n",
+                dbgmsg);
+
+        //pc, cpsr の値を取っておく
+        if (getCPSR().getTBit()) {
+            //Thumb モード
+            pcOrg = getRegRaw(15) + 2;
+        } else {
+            //ARM モード
+            pcOrg = getRegRaw(15) + 4;
+        }
+        cpsrOrg = getCPSR().getValue();
+
+        //未定義モード、ARM 状態、割り込み禁止、
+        //へ移行
+        getCPSR().setMode(PSR.MODE_UND);
+        getCPSR().setTBit(false);
+        //F flag is not affected
+        getCPSR().setIBit(true);
+
+        //lr, spsr に例外前の pc, cpsr を保存する
+        setReg(14, pcOrg);
+        getSPSR().setValue(cpsrOrg);
+
+        //未定義例外ベクタへ
+        if (isHighVector()) {
+            setRegRaw(15, 0xffff0004);
+        } else {
+            setRegRaw(15, 0x00000004);
+        }
+
+        //tentative...
+        //TODO: Not implemented
+        throw new IllegalArgumentException("Sorry, not implemented.");
+    }
+
+    /**
+     * ソフトウェア割り込み例外を発生させます。
+     *
+     * この例外は swi 命令を実行したときに生成されます。
+     *
+     * @param dbgmsg デバッグ用のメッセージ
+     */
+    public void doExceptionSoftware(String dbgmsg) {
+        int pcOrg, cpsrOrg;
+
+        //System.out.printf("Exception: Software interrupt by '%s'.\n",
+        //        dbgmsg);
+
+        //pc, cpsr の値を取っておく
+        if (getCPSR().getTBit()) {
+            //Thumb モード
+            pcOrg = getRegRaw(15) + 2;
+        } else {
+            //ARM モード
+            pcOrg = getRegRaw(15) + 4;
+        }
+        cpsrOrg = getCPSR().getValue();
+
+        //スーパバイザモード、ARM 状態、割り込み禁止、
+        //へ移行
+        getCPSR().setMode(PSR.MODE_SVC);
+        getCPSR().setTBit(false);
+        //F flag is not affected
+        getCPSR().setIBit(true);
+
+        //lr, spsr に例外前の pc, cpsr を保存する
+        setReg(14, pcOrg);
+        getSPSR().setValue(cpsrOrg);
+
+        //ソフトウェア割り込み例外ベクタへ
+        if (isHighVector()) {
+            setRegRaw(15, 0xffff0008);
+        } else {
+            setRegRaw(15, 0x00000008);
+        }
+    }
+
+    /**
+     * プリフェッチアボート例外を発生させます。
+     *
+     * この例外は無効な命令を実行したときに生成されます。
+     *
+     * @param dbgmsg デバッグ用のメッセージ
+     */
+    public void doExceptionPrefetch(String dbgmsg) {
+        int pcOrg, cpsrOrg;
+
+        //System.out.printf("Exception: Prefetch abort by '%s'.\n",
+        //        dbgmsg);
+
+        //pc, cpsr の値を取っておく
+        //Thumb, ARM モード
+        pcOrg = getRegRaw(15) + 4;
+        cpsrOrg = getCPSR().getValue();
+
+        //アボートモード、ARM 状態、割り込み禁止、
+        //へ移行
+        getCPSR().setMode(PSR.MODE_ABT);
+        getCPSR().setTBit(false);
+        //F flag is not affected
+        getCPSR().setIBit(true);
+
+        //lr, spsr に例外前の pc, cpsr を保存する
+        setReg(14, pcOrg);
+        getSPSR().setValue(cpsrOrg);
+
+        //プリフェッチアボート例外ベクタへ
+        if (isHighVector()) {
+            setRegRaw(15, 0xffff000c);
+        } else {
+            setRegRaw(15, 0x0000000c);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 637:656, 674:693, 711:730.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.